src/network/src/esp/src/TCPClient.cpp

```diff
@@ -0,0 +1,58 @@
+#include "TCPClient.h"
+#include "NetworkConfig.h"
+#include "SocketFactory.h"
+#include "lwip/sockets.h"
+
+TCPClient::TCPClient(ILogger& logger) : m_logger(logger) {
+    m_hasSocket = false;
+    m_socketFd = NO_SOCKET;
+}
+
+bool TCPClient::setDestination(const char* address) {
+    if (m_hasSocket) {
+        m_logger.log(LogLevel::Error, "Trying to override socket for TCP client");
+    } else if ((m_socketFd = SocketFactory::createTCPClient(
+                    address, NetworkConfig::getCommunicationPort())) == NO_SOCKET) {
+        m_logger.log(LogLevel::Error, "Failed to acquire connected socket for TCP client");
+    } else {
+        m_hasSocket = true;
+        return true;
+    }
+
+    return false;
+}
+
+bool TCPClient::send(const uint8_t* data, uint16_t length) {
+
+    if (!m_hasSocket) {
+        m_logger.log(LogLevel::Error, "Trying to send message with no socket client socket set");
+        return false;
+    }
+
+    if (lwip_send(m_socketFd, data, length, 0) < 0) {
+        m_logger.log(LogLevel::Error, "Failed to send message");
+    } else {
+        lwip_close(m_socketFd);
+        m_socketFd = NO_SOCKET;
+        m_hasSocket = false;
+        return true;
+    }
+    return false;
+}
+
+bool TCPClient::receive(uint8_t* data, uint16_t length) {
+    (void)data;
+    (void)length;
+    m_logger.log(LogLevel::Error, "Receiving not supported on TCP client");
```

Comment:
Is this a design choice? If we wan't to do bidirectional communication, we will be using two sockets?

Reply:
Yes, though I am not a fan of this myself. The issue arises from the IProtobufStream interface. It assumes that the same class will act as both a serializer and a deserializer. In this case, the esp has a TCP server active at all times to accept inbound communication. To avoid locking the reception of messages, a client socket is created for outbound transmission. This way, the module can receive at all times and also have multiple outbound sockets.

Technically, the tcp socket can both send and receive, but for the client is used as a serializer for protobuf and as such, receive should never be called. A proper fix would be to split the IProtobufStream into two interfaces. Classes that implement the send and receive function would inherit the two interfaces. This is a rework that would affect the HiveMind and ROS-Bridge, that's why I didn't do it in this development, but it could be done in the future.

Reply:
@xgroleau Pretty sure this discussion will interest you

Reply:
I didn't separate the interface in two since that would either require multiple inheritance (the tcp socket would be a output and an input, so it would require to inherit from both), which I try to avoid on embedded or always creating wrapper classes since we generally have read/write comm interface.

What I see you could do is just create your own "only write" and "only read" interface (you already have ones actually, just remove the IProtobuf stream interface and remove the unecessary functions). Then you could create a wrapper class that is an IProtobufStream that takes one "only write" and "only read" at the constructor and call those internally?

Reply:
I think I have an idea for this. It would involve interface composition on the IProtobufStreamInterface that would not require any change on classes already implementing the interface. I'll make a proposal PR on Propolis for this that should resolve this issue.